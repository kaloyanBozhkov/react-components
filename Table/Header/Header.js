import React from 'react'
import PropTypes from 'prop-types'
import Icon from 'UI/Icon/Icon'
import styles from './styles.module.scss'

const Header = ({ pageName, headerCells }) => {

  const classes = [
    styles.tableHeader, 
    styles[pageName] || ''
  ].join(' ').trim()

  return (<header aria-label="table header" className={classes}>
    {headerCells.map(
      ({ label, centered = false, withIcon = false, onClick = undefined }, index) => {
        // apply specific styling based on needs for that table
        const classes = [
          centered ? styles.centered : '',
          withIcon ? styles.withIcon : '',
          onClick ? styles.withAction : ''
        ].join(' ').trim() || undefined

        return (
          <div key={`${label.replace(' ', '')}-${index}`} className={classes} onClick={onClick}>
            <p>{label}</p>
            {withIcon && <Icon icon={withIcon} />}
          </div>
        )
      }
    )}
  </header>)
}

Header.propTypes = {
  pageName: PropTypes.string.isRequired,
  headerCells: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      centered: PropTypes.bool,
      withIcon: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.bool
      ])
    })
  )
}

export default Header
