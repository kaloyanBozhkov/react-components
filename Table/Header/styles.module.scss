@import 'scss/abstracts/_variables';
.tableHeader {
  min-width: 60rem;
  display: grid;
  grid-gap: 1px;
  height: 4.2rem;
  background-color: #eaecef;
  padding: 1px 1px 0px 1px;

  > div {
    display: flex;
    align-items: center;

    font-size: 1rem;
    text-transform: uppercase;

    transition: 300ms;
    background-color: #f2f5f7;

    &.withAction {
      cursor: pointer;
      &:hover {
        background-color: #e5e5e9;
      }
    }

    padding: $medium-padding * 1.5;

    &.centered {
      justify-content: center;
    }

    &.withIcon {
      p {
        padding-right: 2rem;
      }
    }
  }

  //CUSTOM STYLINGS BASED ON CONTENT HERE

  &.accountSummary {
    grid-template-columns: repeat(4, 1fr);
  }
}
