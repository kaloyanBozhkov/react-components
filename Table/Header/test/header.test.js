import React from 'react'
import { shallow } from 'enzyme'

import Header from '../Header'

describe('testing Table component conditional renders for header', () => {

    it('should render header component', () => {
        const headerCells = [
            {
                label: 'columnHeader'
            }
        ]
        
        const wrapper = shallow(<Header headerCells={headerCells} pageName="somePageModifier" />)
        
        expect(wrapper).toMatchSnapshot()
    }) 

    it('should render multiple columns/headerCells with proper label values', () => {
        const headerCells = [
            {
                label: 'columnHeader1'
            },
            {
                label: 'columnHeader2',
            },            
            {
                label: 'columnHeader3',
            },
        ]
        
        const wrapper = shallow(<Header headerCells={headerCells} pageName="somePageModifier" />)

        //3 columns/headerCells
        expect(wrapper.children()).toHaveLength(3)

        //test label content
        expect(wrapper.childAt(0).find('p').text()).toEqual('columnHeader1')
        expect(wrapper.childAt(1).find('p').text()).toEqual('columnHeader2')
        expect(wrapper.childAt(2).find('p').text()).toEqual('columnHeader3')
    })

    it('should render just a header with label, no optional classes', () => {
        const headerCells = [
            {
                label: 'columnHeader'
            }
        ]
        
        const wrapper = shallow(<Header headerCells={headerCells} pageName="somePageModifier" />)
        const headerCell = wrapper.childAt(0)

        //test their unique display related classes
        expect(headerCell.hasClass('centered')).toEqual(false)
        expect(headerCell.hasClass('withIcon')).toEqual(false)
    })

    it('should render just a header with label, and optional class "centered"', () => {
        const headerCells = [
            {
                label: 'columnHeader',
                centered: true
            }
        ]
        
        const wrapper = shallow(<Header headerCells={headerCells} pageName="somePageModifier" />)
        const centeredHeaderCell = wrapper.childAt(0)

        expect(centeredHeaderCell.hasClass('centered')).toEqual(true)
        expect(centeredHeaderCell.hasClass('withIcon')).toEqual(false)
    })

    it('should render just a header with label, and optional class "withIcon"', () => {
        const headerCells = [
            {
                label: 'columnHeader',
                withIcon: true
            }
        ]
        
        const wrapper = shallow(<Header headerCells={headerCells} pageName="somePageModifier" />)
        const withIconHeaderCell = wrapper.childAt(0)

        expect(withIconHeaderCell.hasClass('centered')).toEqual(false)
        expect(withIconHeaderCell.hasClass('withIcon')).toEqual(true)
        expect(withIconHeaderCell.find('Icon')).toHaveLength(1)
    })

    it('should render just a header with label, and optional class "withIcon" and "centered', () => {
        const headerCells = [
            {
                label: 'columnHeader',
                centered: true,
                withIcon: 'sort'
            },    
        ]
        
        const wrapper = shallow(<Header headerCells={headerCells} pageName="somePageModifier" />)
        const headerCell = wrapper.childAt(0)

        expect(headerCell.hasClass('centered')).toEqual(true)
        expect(headerCell.hasClass('withIcon')).toEqual(true)
    })

    it('should render jsut a header with label, that is clickable and can run func', () => {
        const mockFunc = jest.fn(() => "works")

        const headerCells = [
            {
                label: 'columnHeader',
                onClick: mockFunc
            }
        ]
        
        const wrapper = shallow(<Header headerCells={headerCells} pageName="somePageModifier" />)
        const headerCell = wrapper.childAt(0)
        headerCell.simulate('click')
        
        expect(mockFunc).toHaveBeenCalled()
    })
})