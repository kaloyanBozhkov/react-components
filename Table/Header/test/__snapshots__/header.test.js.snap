// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Table component conditional renders for header should render header component 1`] = `
<header
  aria-label="table header"
  className="tableHeader somePageModifier"
>
  <div
    key="columnHeader-0"
  >
    <p>
      columnHeader
    </p>
  </div>
</header>
`;
